platform/macos/os_macos.mm

```diff
@@ -47,17 +47,69 @@
 #include <sys/sysctl.h>
 
 void OS_MacOS::pre_wait_observer_cb(CFRunLoopObserverRef p_observer, CFRunLoopActivity p_activiy, void *p_context) {
-	// Prevent main loop from sleeping and redraw window during modal popup display.
-	// Do not redraw when rendering is done from the separate thread, it will conflict with the OpenGL context updates.
-
-	DisplayServerMacOS *ds = (DisplayServerMacOS *)DisplayServer::get_singleton();
-	if (get_singleton()->get_main_loop() && ds && !get_singleton()->is_separate_thread_rendering_enabled() && !ds->get_is_resizing()) {
-		Main::force_redraw();
-		if (!Main::is_iterating()) { // Avoid cyclic loop.
-			Main::iteration();
+	OS_MacOS *os = static_cast<OS_MacOS *>(OS::get_singleton());
+
+	if (os->run_stage == APP_RUN_STAGE_SETUP) {
+		Error err;
+		@autoreleasepool {
+			err = Main::setup(os->execpath, os->argc, os->argv);
+		}
+
+		if (err == OK) {
+			int ret;
+			@autoreleasepool {
+				ret = Main::start();
+			}
+			if (ret == EXIT_SUCCESS) {
+				if (os->main_loop) {
+					os->run_stage = APP_RUN_STAGE_RUN;
+					@autoreleasepool {
+						os->main_loop->initialize();
+					}
+				} else {
+					os->run_stage = APP_RUN_STAGE_END_NO_CLEANUP;
+					os->should_terminate = true;
+				}
+			} else {
+				os->set_exit_code(EXIT_FAILURE);
+				os->run_stage = APP_RUN_STAGE_END_NO_CLEANUP;
+				os->should_terminate = true;
+			}
+		} else {
+			if (err == ERR_HELP) { // Returned by --help and --version, so success.
+				os->set_exit_code(EXIT_SUCCESS);
+			} else {
+				os->set_exit_code(EXIT_FAILURE);
+			}
+			os->run_stage = APP_RUN_STAGE_END_NO_CLEANUP;
+			os->should_terminate = true;
+		}
+	}
+
+	if (os->run_stage == APP_RUN_STAGE_RUN) {
+		@autoreleasepool {
+			@try {
+				if (DisplayServer::get_singleton()) {
+					DisplayServer::get_singleton()->process_events(); // Get rid of pending events.
+				}
+				os->joypad_apple->process_joypads();
+
+				if (Main::iteration()) {
+					os->run_stage = APP_RUN_STAGE_END;
+					os->should_terminate = true;
+				}
+			} @catch (NSException *exception) {
+				ERR_PRINT("NSException: " + String::utf8([exception reason].UTF8String));
+			}
 		}
 	}
 
+	if (os->should_terminate) {
+		CFRunLoopRemoveObserver(CFRunLoopGetCurrent(), static_cast<OS_MacOS *>(OS::get_singleton())->pre_wait_observer, kCFRunLoopCommonModes);
+		[NSApp stop:nil];
+		[NSApp abortModal];
+	}
```

Comment:
With the other changes, you can then move this logic to the [`applicationWillTerminate:`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/applicationwillterminate(_:)?language=objc) delegate method, which runs during termination. You won't need to call the other `[NSApp ..]` methods here, as it is already shutting down.

Reply:
Since `terminate` is not used, this won't always be called.